mautic/mautic
# Api utmTags enhancements

| Q  | A
| --- | ---
| Bug fix? | Yes
| New feature? | Yes
| Related user documentation PR URL | 
| Related developer documentation PR URL |  https://github.com/mautic/developer-documentation/pull/47
| Issues addressed (#s or URLs) | 
| BC breaks? | 
| Deprecations? | 

#### Description:
This PR addresses issues with UTM Tags via the API. Currently the field name is included in the returned contact data but is always empty.

This first commit addresses that issue. After applying this PR a contact/get will include UTM Tag data that has been recorded.

***Update 1***
Now we have endpoints to add and remove utmTag sets. To add utmTags to an existing contact (313) you would use the following endpoint. `/api/contacts/313/utm/add`.

To remove the utmTag set use: `/api/contacts/313/utm/1234/remove`

The following fields are available:
```
utm_campaign
utm_source 
utm_medium 
utm_content
utm_term 
useragent
url
referer
query
remotehost
lastActive
```
If you supply the last active date, the contact Last Active date is also updated. 
If you provide a Useragent, a new Device will be assigned to the contact. 

#### Steps to test this PR:
1. Install this PR
2. **clear the cache**
3. Add a contact and record the ID.
4. create a template file for cURL something similar to the following 
```
lastActive=2017-01-17T00%3A30%3A08%2B00%3A00&
utm_campaign=apicamp&
utm_source=fb&
utm_medium=social&
utm_content=fbad&
utm_term=mautic+api&
url=/product/fbad01/&
remotehost=example.com&
query%5Bprodid%5D=bar&
query%5Brefid%5D=andy&
useragent=Mozilla%2F5.0+%28iPhone%3B+CPU+iPhone+OS+10_0+like+Mac+OS+X%29+AppleWebKit%2F602.1.50+%28KHTML%2C+like+Gecko%29+Mobile%2F14A346+%5BFBAN%2FFBIOS%3BFBAV%2F68.0.0.49.70%3BFBBV%2F41924288%3BFBRV%2F0%3BFBDV%2FiPhone8%2C1%3BFBMD%2FiPhone%3BFBSN%2FiOS%3BFBSV%2F10.0%3BFBSS%2F2%3BFBCR%2FVerizon%3BFBID%2Fphone%3BFBLC%2Fen_US%3BFBOP%2F5
```
5. Ensure you know the appropriate Basic Auth string and run the following. (assuming you saved the template as `utm1.txt`)
```
curl -v -H "Authorization: basic YXBpdXNlcjpwYXNzd29yZDE=" -d @utm1.txt http://mautic-instance.dev/api/contacts/313/utm/add
```
6. The returned json includes the ID for the UtmTags added. Note the ID and perform the following to remove it.
```
curl -v -H "Authorization: basic YXBpdXNlcjpwYXNzd29yZDE=" -d "" http://mautic-instance.dev/api/contacts/313/utm/1234/remove
```
#### Steps to reproduce the bug:
1. See above
